File: exercism/python/minesweeper/minesweeper.py
SPACE, MINE, NS_EDGE, EW_EDGE, CORNER = ' *|-+'

product = lambda xs, ys: [(x, y) for x in xs for y in ys]
OFFSETS = product((-1, 0, 1), (-1, 0, 1))

chunk = lambda n, xs: [xs[i*n:(i+1)*n] for i in range(len(xs) // n)]


def board(M):
    validate(M)
    width, height = (len(M[0]), len(M))
    syms = [counts(M, i, j) for i, j in product(range(height), range(width))]
    return [''.join(row) for row in chunk(width, syms)]


def counts(M, i, j):
    if M[i][j] != SPACE: return M[i][j]
    return str(sum(M[i + di][j + dj] == MINE for di, dj in OFFSETS) or ' ')


def validate(M):
    MT = list(zip(*M))
    edges = M[0] + M[-1] + ''.join(MT[0] + MT[-1])
    pieces = ''.join(row[1:-1] for row in M[1:-1])

    errors = [msg for msg, predicate in (
        ('Irregular board shape', len(set(map(len, M))) > 2),
        ('Weird edge', set(edges) - set(NS_EDGE + EW_EDGE + CORNER)),
        ('Weird piece', set(pieces) - set(SPACE + MINE))
    ) if predicate]
    if errors: raise ValueError(errors[0])

File: exercism/go/minesweeper/minesweeper.go
package minesweeper

import "errors"

type delta struct{ i, j int }

var deltas = []delta{{-1, -1}, {0, -1}, {1, -1}, {-1, 0}, {1, 0}, {-1, 1}, {0, 1}, {1, 1}}
var invalidBoard = errors.New("Invalid board")

func (b *Board) Count() error {
	height, width := len(*b), 0
	for i, row := range *b {
		if i == 0 {
			width = len(row)
		}
		for j, x := range row {
			switch {
			case len(row) != width:
				return invalidBoard
			case i == 0 && j == 0,
				i == 0 && j == width-1,
				i == height-1 && j == 0,
				i == height-1 && j == width-1:
				if x != '+' {
					return invalidBoard
				}
			case i == 0,
				i == height-1:
				if x != '-' {
					return invalidBoard
				}
			case j == 0,
				j == width-1:
				if x != '|' {
					return invalidBoard
				}
			case x == '*':
				continue
			case x == ' ':
				var count byte = '0'
				for _, d := range deltas {
					if (*b)[i+d.i][j+d.j] == '*' {
						count += 1
					}
				}
				if count > '0' {
					row[j] = count
				}
			default:
				return invalidBoard
			}
		}
	}
	return nil
}
